### PR TITLE
add prediction_gradient method for MLPs

### PR DESCRIPTION
**Pull request by mheilman**
This adds a method to facilitate the interpretation of MLP model predictions.  Specifically, it enables one to compute the gradient of the final output layer (pre-sigmoid logits for logistic regression, etc.) with respect to the input features.

I'm leaving sparse inputs to a future PR.  Also, I have some notebooks illustrating the method on toy datasets that might be worth adding.

### CODE REVIEW

**Comment by mheilman** [comment]
@hinnefe2, any interest in reviewing this?

**Comment by hinnefe2** [comment]
Sure thing

File: muffnn/mlp/base.py [hunk 1]
```diff
@@ -301,6 +301,26 @@ def _set_up_graph(self):
...
+        # We use the top layer before the output activation function
+        # (e.g., softmax, sigmoid) following
+        # https://arxiv.org/pdf/1312.6034.pdf
+        if self.is_sparse_:
+            self._prediction_gradient = None
+        else:
+            output_shape = self.output_layer_.get_shape()
+            if len(output_shape) == 1:
+                self._prediction_gradient = tf.gradients(
+                    t, self._input_values)[0]
```
Comment (hinnefe2) [inline muffnn/mlp/base.py]: I'm a little confused by the `[0]` indexing here - why do we only want the first element?

File: muffnn/mlp/base.py [hunk 2]
```diff
@@ -301,6 +301,26 @@ def _set_up_graph(self):
...
+            else:
```
Comment (hinnefe2) [inline muffnn/mlp/base.py]: Bit of a stretch, but should this check for `len(output_shape) == 2`? The only case I can think of where it might matter is if you had a timeseries multiclass problem though so maybe not necessary.

File: muffnn/mlp/base.py [hunk 3]
```diff
@@ -301,6 +301,26 @@ def _set_up_graph(self):
...
+                # According to the tf.gradients documentation, it looks like
+                # we have to compute gradients separately for each output
+                # dimension and then stack them for multiclass/label data.
+                self._prediction_gradient = tf.stack([
+                    tf.gradients(t[:, i], self._input_values)[0]
+                    for i in range(self.output_layer_.get_shape()[1])
```
Comment (hinnefe2) [inline muffnn/mlp/base.py]: (nitpick) Since we already defined `output_shape = self.output_layer_.get_shape()` this line might read a little easier if we used that variable.

File: muffnn/mlp/tests/test_mlp_classifier.py [hunk 4]
```diff
@@ -357,3 +357,44 @@ def test_sample_weight(make_dataset_func, dataset_kwargs):
...
+
+
+def test_prediction_gradient():
+    """Test computation of prediction gradients."""
+    # Binary classification
+    n_classes = 1
+    mlp = MLPClassifier(n_epochs=100, random_state=42, hidden_units=(5,))
+    X, y = make_classification(
+        n_samples=1000, n_informative=n_classes, n_redundant=0,
+        n_classes=n_classes, n_clusters_per_class=1, shuffle=False)
```
Comment (hinnefe2) [inline muffnn/mlp/tests/test_mlp_classifier.py]: An explicit `n_features=20` might be helpful here. Before I checked the docs I assumed `n_features = n_informative + n_redundant` and was confused.

File: muffnn/mlp/tests/test_mlp_classifier.py [hunk 5]
```diff
@@ -357,3 +357,44 @@ def test_sample_weight(make_dataset_func, dataset_kwargs):
...
+    mlp.fit(X, y)
+    grad = mlp.prediction_gradient(X)
+    grad_means = grad.mean(axis=0)
+    assert grad.shape == X.shape
+    # Check that only the informative feature has a large gradient.
+    assert np.abs(grad_means[0]) > 1.
+    for m in grad_means[1:]:
+        assert np.abs(m) < 0.5
```
Comment (hinnefe2) [inline muffnn/mlp/tests/test_mlp_classifier.py]: Why 0.5?

File: muffnn/mlp/tests/test_mlp_regressor.py [hunk 6]
```diff
@@ -139,3 +140,24 @@ def test_embedding_specific_layer():
...
+        n_samples=1000, n_features=10, n_informative=1, shuffle=False)
+    mlp.fit(X, y)
+    grad = mlp.prediction_gradient(X)
+    grad_means = grad.mean(axis=0)
+    assert grad.shape == X.shape
+    # Check that only the informative feature has a large gradient.
+    assert np.abs(grad_means[0]) > 0.5
+    for m in grad_means[1:]:
+        assert np.abs(m) < 0.1
+
```
Comment (hinnefe2) [inline muffnn/mlp/tests/test_mlp_regressor.py]: It looks like `make_regression` can handle multiclass problems (with `n_targets`) - maybe add the multiclass test from the MLPClassifier here too?

Comment (mheilman) [reply muffnn/mlp/tests/test_mlp_regressor.py]: MLPRegressor doesn't do multilabel yet, unfortunately (#85)

File: muffnn/mlp/tests/test_mlp_classifier.py (hunk 5 above)
Comment (mheilman) [reply muffnn/mlp/tests/test_mlp_classifier.py]: This was fairly arbitrary.  Let me write a comment here and see if you think it makes snse

File: muffnn/mlp/base.py (hunk 3 above)
Comment (mheilman) [reply muffnn/mlp/base.py]: oh, good catch, thanks

File: muffnn/mlp/base.py (hunk 2 above)
Comment (mheilman) [reply muffnn/mlp/base.py]: I don't think it's possible to have dimension > 2, but an extra check and valueerror wouldn't hurt. will add

**Comment by mheilman** [comment]
> What do you think of a warning if this is called on data with large variation in feature ranges...

I'd prefer not to add a warning since it might be hard to get that right, and it might be annoying.  Also, I wouldn't want somebody to think the lack of a warning means that things are perfect.  How about a comment to the docstring?

File: muffnn/mlp/base.py (hunk 1 above)
Comment (mheilman) [reply muffnn/mlp/base.py]: `tf.gradients` returns a list of gradients dy/dx, one per given x.  In this case, that has length 1.  Also, it's polymorphic, so `tf.gradients([y], [x]) == tf.gradients(y, x)`.

```
(Pdb) tf.gradients(t, [self._input_values])
[<tf.Tensor 'gradients/layer_0/MatMul_grad/MatMul:0' shape=(?, 20) dtype=float32>]
(Pdb) tf.gradients(t, self._input_values)
[<tf.Tensor 'gradients_1/layer_0/MatMul_grad/MatMul:0' shape=(?, 20) dtype=float32>]
```

Comment (hinnefe2) [reply muffnn/mlp/base.py]: Sorry, still a bit confused here. The docs say that gradients returns "A list of sum(dy/dx) for each x in xs." Is the sum "for each x in xs" over the columns in a given instance or over instances?

The first option (sum over columns) doesn't make sense to me because then you don't get per-column gradients, i.e. feature importances. The second option doesn't make sense with the indexing because it seems like `gradients[0]` would just give you dy/dx_0.

For the non-multiclass case I'm assuming:
```
t.shape = (n_examples,)
self._input_values.shape = (n_examples, n_features)

tf.gradients().shape = (n_examples,)  # sum over columns?
tf.gradients().shape = (n_features,)  # sum over instances?
```
Where am I going wrong here?

**Comment by beckermr** [comment]
Let's not coddle our users. They have to know what they are doing no matter how many warnings we throw at them. :)

File: muffnn/mlp/base.py (hunk 1 above)
Comment (mheilman) [reply muffnn/mlp/base.py]: Perhaps this will help (sorry this is tricky and I didn't explain it well)

```python
# Here's an illustration of tf.gradients behavior for a linear transformation.
import numpy as np
import tensorflow as tf
# n-by-3
x = tf.placeholder(dtype=tf.float32, shape=(None, 3))
# 3-by-2
w_np = np.array([[1, 2],
                 [3, 4],
                 [5, 6]], dtype=np.float32)
w = tf.constant(w_np)
# n-by-2
y = tf.matmul(x, w)
print("output shape:", y.get_shape())
# n-by-3
g_multioutput = tf.gradients(y, x)
g_singleoutput = tf.gradients(y[:,0], x)
session = tf.Session()
x_np = np.zeros((10, 3), dtype=np.float32)
result_singleoutput = session.run(g_singleoutput, feed_dict={x: x_np})
print("dy[0]/dx:\n", result_singleoutput, sep="")
result_multioutput = session.run(g_multioutput, feed_dict={x: x_np})
print("dy/dx:\n", result_multioutput, sep="")
# The gradient is the same for each row of input because this is just linear.
# The gradient will just be the column of the linear transformation w,
# or the sum across columns for the multioutput case since tf.gradients
# will sum across the columns of y.
for row_single, row_multi in zip(result_singleoutput, result_multioutput):
    assert np.allclose(row_single, w_np[:, 0])
    assert np.allclose(row_multi, w_np.sum(axis=1))
```

Comment (hinnefe2) [reply muffnn/mlp/base.py]: Ok I think I get this now - the `[0]` is because we get back a length - 1 list, but the element in that list is the array of n_examples x n_features with the gradients for each feature for each instance. The `tf.stack` business is because we don't want to sum over y dimensions in the multiclass case (in case a gradient is + for some dv but - for another).

I think my confusion was that you get back a n_features x n_gradients array wrapped in a list - I thought you would just get back the array, which is more of an issue with my reading of the tf docs. When I read "for each `x` in `xs`" my first interpretation is to think of the single `x` as a row, not as a tensor.
